lightningd/gossip_control.c

```diff
@@ -614,6 +615,167 @@ static const struct json_command listprivateinbound_command = {
+		/* FIXME: check the status of these calls */
+		json_to_node_id(buffer, id_tok, &node_id);
+		json_to_node_id(buffer, id_tok, &inbound_node.id);
+		const jsmntok_t *cap_tok;
+		struct amount_msat capacity_msat;
+		cap_tok = json_get_member(buffer, t, "capacity_msat");
+		json_to_msat(buffer, cap_tok, &capacity_msat);
+		json_to_msat(buffer, cap_tok, &inbound_node.incoming_capacity_msat);
+		// Add this json object later instead - just populate data here.
+		// json_object_start(js, NULL);
+		// json_add_node_id(js, "id", &node_id);
+		// json_add_amount_msat(js, "capacity_msat", capacity_msat);
+		const jsmntok_t *feats;
+		feats = json_get_member(buffer, t, "peer_features");
+		inbound_node.features = json_strdup(cmd, buffer, feats);
+		// json_object_end(js);
+		tal_arr_expand(&inbound_nodes, inbound_node);
+		const jsmntok_t *chan_tok;
+		chan_tok = json_get_member(buffer, t, "inbound_channel");
+		if (!chan_tok)
+			continue;
+		/* FIXME: this should be a list? - iterate through */
+		struct inbound_channel inbound_chan;
+		/* FIXME: extract fields and populate inbound_chan */
+		const jsmntok_t *d;
+		inbound_chan.id = node_id;
+		d = json_get_member(buffer, chan_tok, "htlc_max_msat");
+		json_to_msat(buffer, d, &inbound_chan.htlc_max_msat);
+		// inbound_chan.features = json_strdup(cmd, buffer, d);
+		tal_arr_expand(&inbound_channels, inbound_chan);
+	}
+	struct peer_node_id_map_iter it;
+	struct peer *peer;
+	for (peer = peer_node_id_map_first(cmd->ld->peers, &it);
+	     peer;
+	     peer = peer_node_id_map_next(cmd->ld->peers, &it)) {
+		struct channel *c;
+		list_for_each(&peer->channels, c, list) {
+			/* Pull channel info from inbound_channels if found,
+			 * otherwise see if there's a private update. */
+			if (!c->private_update)
+				continue;
+			json_object_start(js, NULL);
+			json_add_short_channel_id(js,
+						  "short_channel_id",
+						  &c->private_update->scid);
+			json_add_u32(js, "fee_base",
+				     c->private_update->fee_base);
+			json_add_u32(js, "fee_ppm",
+				     c->private_update->fee_ppm);
+			json_add_u32(js, "cltv_delta",
+				     c->private_update->cltv_delta);
+			json_add_amount_msat(js, "htlc_minimum_msat",
+					     c->private_update->htlc_minimum_msat);
+			json_add_amount_msat(js, "htlc_maximum_msat",
+					     c->private_update->htlc_maximum_msat);
+			/* FIXME: calculate incoming balance and constrain here.
+			 * Also apply node total inbound capacity. */
+			json_add_amount_msat(js, "incoming_capacity_msat",
+					     c->private_update->htlc_maximum_msat);	
+			json_object_end(js);
+		}
+	}
+
+	/* iterate through list of channels, constrain by topo data */
+	json_array_end(js);
+	tal_free(inbound_nodes);
+	tal_free(inbound_channels);
+	was_pending(command_success(cmd, js));
+}
+
+static struct command_result *json_listincoming2(struct command *cmd,
+						 const char *buffer,
+						 const jsmntok_t *obj UNNEEDED,
+						 const jsmntok_t *params)
+{
+	struct peer *peer;
```

Comment:
listincoming2 belongs in topology.

1. It needs to use listpeerchannels to get info about private channels/updates.
2. Merge that with its topology info (as it does in getremoteliquidity).
3. Present a unified view of all peers' calculated capacity.